drivers/d3d12/rendering_device_d3d12.cpp

```diff
@@ -0,0 +1,10113 @@
+
+			for (uint32_t y = 0; y < height; y += region_size) {
+				for (uint32_t x = 0; x < width; x += region_size) {
+					uint32_t region_w = MIN(region_size, width - x);
+					uint32_t region_h = MIN(region_size, height - y);
+
+					uint32_t pixel_size = get_image_format_pixel_size(p_texture->format);
+					uint32_t block_w, block_h;
+					get_compressed_image_format_block_dimensions(p_texture->format, block_w, block_h);
+
+					uint32_t region_pitch = (region_w * pixel_size * block_w) >> get_compressed_image_format_pixel_rshift(p_texture->format);
+					region_pitch = ALIGN(region_pitch, D3D12_TEXTURE_DATA_PITCH_ALIGNMENT);
+					uint32_t to_allocate = region_pitch * region_h;
+
+					uint32_t alloc_offset, alloc_size;
+					Error err = _staging_buffer_allocate(to_allocate, required_align, alloc_offset, alloc_size, false);
+					ERR_FAIL_COND_V(err, ERR_CANT_CREATE);
+
+					uint8_t *write_ptr;
+
+					{ // Map.
+						void *data_ptr = nullptr;
+						HRESULT res = staging_buffer_blocks[staging_buffer_current].resource->Map(0, &VOID_RANGE, &data_ptr);
+						ERR_FAIL_COND_V_MSG(res, ERR_CANT_CREATE, "Map failed with error " + vformat("0x%08x", res) + ".");
+						write_ptr = (uint8_t *)data_ptr;
+						write_ptr += alloc_offset;
+					}
+
+					ERR_FAIL_COND_V(region_w % block_w, ERR_BUG);
+					ERR_FAIL_COND_V(region_pitch % block_w, ERR_BUG);
+					ERR_FAIL_COND_V(region_h % block_h, ERR_BUG);
+
+					if (block_w != 1 || block_h != 1) {
+						// Compressed image (functions).
+						// Must copy a block region.
+
+						uint32_t block_size = get_compressed_image_format_block_byte_size(p_texture->format);
+						// Re-create current variables in blocky format.
+						uint32_t xb = x / block_w;
+						uint32_t yb = y / block_h;
+						uint32_t wb = width / block_w;
+						// Uint32_t hb = height / block_h;.
+						uint32_t region_wb = region_w / block_w;
+						uint32_t region_hb = region_h / block_h;
+						_copy_region(read_ptr, write_ptr, xb, yb, region_wb, region_hb, wb, region_pitch, block_size);
+					} else {
+						// Regular image (pixels).
+						// Must copy a pixel region.
+						_copy_region(read_ptr, write_ptr, x, y, region_w, region_h, width, region_pitch, pixel_size);
+					}
+
+					{ // Unmap.
+						staging_buffer_blocks[staging_buffer_current].resource->Unmap(0, &VOID_RANGE);
+					}
+
+					D3D12_PLACED_SUBRESOURCE_FOOTPRINT src_footprint = {};
+					src_footprint.Offset = alloc_offset;
+					src_footprint.Footprint = CD3DX12_SUBRESOURCE_FOOTPRINT(
+							d3d12_formats[p_texture->format].family,
+							region_w,
+							region_h,
+							1,
+							region_pitch);
+					CD3DX12_TEXTURE_COPY_LOCATION copy_src(staging_buffer_blocks[staging_buffer_current].resource, src_footprint);
+
+					CD3DX12_BOX src_box(0, 0, region_w, region_h);
+					p_command_list->CopyTextureRegion(&copy_dst, x, y, z, &copy_src, &src_box);
+
+					staging_buffer_blocks.write[staging_buffer_current].fill_amount += alloc_size;
```

Comment:
```suggestion
					staging_buffer_blocks.write[staging_buffer_current].fill_amount = alloc_offset + alloc_size;
```

The current code does not take into account any alignment that may be applied to alloc_offset which will result in overlapping blocks. This can be triggered by having a 223x399 sprite where the very bottom row of pixels will become the top row of the next block to the right. The technique shown in this suggestion comes from `RenderingDeviceD3D12::_buffer_update()`. Technically this bug also exists in the Vulkan code (as it's identical) but it isn't triggered due to the less strict alignment rules.

Reply:
Good catch! Fixed.

Reply:
Your commit seems to have left the `+=` in there.